OpenRA.Game/GameRules/WeaponInfo.cs

```diff
@@ -219,11 +219,13 @@ public void Impact(Target target, Actor firedBy)
 			var args = new WarheadArgs
 			{
 				Weapon = this,
-				Source = firedBy.CenterPosition,
 				SourceActor = firedBy,
 				WeaponTarget = target
 			};
 
+			if (firedBy.OccupiesSpace != null)
+				args.Source = firedBy.CenterPosition;
```

Comment:
I suspect that leaving `Source` with its default value (I guess `WPos.Zero`) may cause other bugs. We probably want to fall back to `target.CenterPosition` instead.

Reply:
You're right. I'm not yet sure if defaulting to the target center is better than making this a nullable. (With a nullable the receiver of the args immediately knows that there is no source position. Falling back to the target position will make it harder to detect if there is no real source.)
`Source` doesn't appear to be used in our code (yet), so maybe @reaperrr can explain what the plan for it is and what is best to use here. (I.e. would any code using this actually want to have a special case for when there is no real source?)

Reply:
Making it nullable is a good idea.

Reply:
I made it nullable for now.

Reply:
`Source` is in some cases used to calculate the impact direction in my downstream implementation of directional armor (might try submitting it upstream later), because calculating that from the warhead impact isn't always feasible (lasers, railguns, area beams, InstantHit).

For my use case, falling back to target.CenterPosition would have worked, too.
But I guess we can leave it nullable and I'll (try to) adapt my code.

Reply:
You'll be able to keep the same behaviour using `args.Source ?? args.WeaponTarget.CenterPosition`.